case-lib/pipeline.sh

```diff
@@ -2,6 +2,19 @@
 
 func_pipeline_export()
 {
+    # For legacy HDA platform, there is no topology, we have to exporting pipeline
+    # information from proc file system.
+    is_sof_used || {
+        filter_str="$@"["$#"]
```

Comment:
In English this almost looks like what people type when cursing....
Again explaining what this is supposed to do wouldn't hurt.

Reply:
> In English this almost looks like what people type when cursing....

I couldnt help laughing at this

Reply:
> In English this almost looks like what people type when cursing....

You should be cursing when quoting, this is by design:

http://rg1-teaching.mpi-inf.mpg.de/unixffb-ss98/quoting-guide.html

Reply:
I still don't understand the purpose of `filter_str="$@"["$#"]`. shellcheck doesn't understand either BTW. 

I just tested  `"$@"["$#"]` with `f() { printf '%s__' "$@"["$#"] ; }` to be sure and this is what it does, it turns:
`"a b"   "c d"   "e f"` into: `"a b"   "c d"   "e f[3]"`.

Is this really what you're trying to do here and why?

Why isn't the following enough? It would match the documentation you just added:

```shell
func_pipeline_export()
{
   local tplg="$1  filter="$2"
  
```

Reply:
OK, I think I finally guessed what you're trying to do here. You're probably trying to define `filter` as the last element of the array, in other words you're trying to do this:
```shell
last_arg()
{
    array_from_args=("$@");
    echo "${array_from_args[${#array_from_args[@]}-1]}"
}
```
If I guessed correctly and taking the last function argument is really what you're trying then:
- why does the function header pretend you're using `$2`
- why silently ignore all the arguments between the first and the last?

Reply:
or "${@: -1}"? if we can assume the filter string will always be the last argument

Reply:
> or "${@: -1}"?

No, that defaults to '1' when "$@" is empty (zero argument)

Reply:
@aiChaoSONG An inline comment would solve the issue.
I think what you want is 
`# to get the last argument only.`